Conflict with RenderLib

(Note: This text has been translated using a translation tool.)

In Minecraft 1.12.2 using Forge 14.23.5.2860, when only adding RenderLib, Universal Tweaks, and its prerequisite mods, the game crashes shortly after a popup window appears. The crash log is provided below:
```
---- Minecraft Crash Report ----

WARNING: coremods are present:
  UniversalTweaksCore ([通用修改] UniversalTweaks-1.12.2-1.14.0.jar)
  MixinBooter (!mixinbooter-10.5.jar)
  ConfigAnytimePlugin (!configanytime-3.0.jar)
  RenderLibPlugin (RenderLib-1.12.2-1.4.1.jar)
Contact their authors BEFORE contacting forge

// You're mean.

Time: 3/9/25 1:26 PM
Description: Initializing game

java.lang.NoClassDefFoundError: meldexun/renderlib/RenderLib
	at meldexun.renderlib.opengl.debug.OpenGLDebugMode.setupDebugOutput(OpenGLDebugMode.java:342)
	at net.minecraft.client.Minecraft.handler$zzg000$renderlib$init_createDisplay(Minecraft.java:3367)
	at net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:458)
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:378)
	at net.minecraft.client.main.Main.main(SourceFile:123)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at oolloo.jlw.Wrapper.invokeMain(Wrapper.java:112)
	at oolloo.jlw.Wrapper.main(Wrapper.java:105)
Caused by: java.lang.ClassNotFoundException: meldexun.renderlib.RenderLib
	at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:101)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 17 more

No Mixin Metadata is found in the Stacktrace.


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Client thread
Stacktrace:
	at meldexun.renderlib.opengl.debug.OpenGLDebugMode.setupDebugOutput(OpenGLDebugMode.java:342)
	at net.minecraft.client.Minecraft.handler$zzg000$renderlib$init_createDisplay(Minecraft.java:3367)
	at net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:458)

-- Initialization --
Details:
Stacktrace:
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:378)
	at net.minecraft.client.main.Main.main(SourceFile:123)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at oolloo.jlw.Wrapper.invokeMain(Wrapper.java:112)
	at oolloo.jlw.Wrapper.main(Wrapper.java:105)

-- System Details --
Details:
	Minecraft Version: 1.12.2
	Operating System: Windows 10 (amd64) version 10.0
	Java Version: 1.8.0_302, GraalVM Community
	Java VM Version: OpenJDK 64-Bit Server VM GraalVM CE 21.2.0 (mixed mode), GraalVM Community
	Memory: 356895992 bytes (340 MB) / 562036736 bytes (536 MB) up to 2959081472 bytes (2822 MB)
	JVM Flags: 3 total; -XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump -Xmn445m -Xmx2969m
	IntCache: cache: 0, tcache: 0, allocated: 0, tallocated: 0
	FML: 
	Loaded coremods (and transformers): 
UniversalTweaksCore ([通用修改] UniversalTweaks-1.12.2-1.14.0.jar)
  
MixinBooter (!mixinbooter-10.5.jar)
  
ConfigAnytimePlugin (!configanytime-3.0.jar)
  
RenderLibPlugin (RenderLib-1.12.2-1.4.1.jar)
  
	Launched Version: 	
	LWJGL: 2.9.4
	OpenGL: NVIDIA GeForce RTX 4060 Laptop GPU/PCIe/SSE2 GL version 4.6.0 NVIDIA 572.70, NVIDIA Corporation
	GL Caps: 
	Using VBOs: Yes
	Is Modded: Definitely; Client brand changed to 'fml,forge'
	Type: Client (map_client.txt)
	Resource Packs: 
	Current Language: ~~ERROR~~ NullPointerException: null
	Profiler Position: N/A (disabled)
	CPU: <unknown>
```

--- a/src/main/java/mod/acgaming/universaltweaks/core/Coremods.java
+++ b/src/main/java/mod/acgaming/universaltweaks/core/Coremods.java
@@ -1,0 +1,41 @@
+package mod.acgaming.universaltweaks.core;
+
+import zone.rong.mixinbooter.Context;
+
+public enum Coremods
+{
+    CHUNKGEN_LIMITER("chunkgenlimit"),
+    OPENMODS("openmods"),
+    OPTIFINE("optifine"),
+    RANDOM_PATCHES("randompatches"),
+    RENDERLIB("renderlib"),
+    SPONGEFORGE("spongeforge"),
+    SURGE("surge"),
+    ;
+
+    private static boolean initialized = false;
+
+    public final String modId;
+    private boolean isLoaded;
+
+    Coremods(String modId)
+    {
+        this.modId = modId;
+        this.isLoaded = false;
+    }
+
+    public static void initFromContext(Context context)
+    {
+        if (initialized) return;
+        initialized = true;
+        for (Coremods coreMod : Coremods.values())
+        {
+            coreMod.isLoaded = context.isModPresent(coreMod.modId);
+        }
+    }
+
+    public boolean isLoaded()
+    {
+        return this.isLoaded;
+    }
+}

--- a/src/main/java/mod/acgaming/universaltweaks/core/UTLoadingPlugin.java
+++ b/src/main/java/mod/acgaming/universaltweaks/core/UTLoadingPlugin.java
@@ -4,16 +4,19 @@ import java.util.*;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import com.google.common.collect.ImmutableMap;
 import net.minecraftforge.common.ForgeVersion;
 import net.minecraftforge.fml.relauncher.FMLLaunchHandler;
 import net.minecraftforge.fml.relauncher.IFMLLoadingPlugin;
 
+import mod.acgaming.universaltweaks.Tags;
 import mod.acgaming.universaltweaks.UniversalTweaks;
 import mod.acgaming.universaltweaks.config.UTConfigBugfixes;
 import mod.acgaming.universaltweaks.config.UTConfigGeneral;
 import mod.acgaming.universaltweaks.config.UTConfigTweaks;
-import mod.acgaming.universaltweaks.util.UTReflectionUtil;
+import zone.rong.mixinbooter.Context;
 import zone.rong.mixinbooter.IEarlyMixinLoader;
 
 @IFMLLoadingPlugin.Name("UniversalTweaksCore")
@@ -21,16 +24,9 @@ import zone.rong.mixinbooter.IEarlyMixinLoader;
 @IFMLLoadingPlugin.SortingIndex(Integer.MIN_VALUE)
 public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
 {
+    // Use this logger to avoid loading the main @Mod class just to be safe.
+    public static final Logger LOGGER = LogManager.getLogger(Tags.MOD_NAME);
     public static final boolean isClient = FMLLaunchHandler.side().isClient();
-    public static final boolean isDev = FMLLaunchHandler.isDeobfuscatedEnvironment();
-
-    public static boolean openModsLoaded;
-    public static boolean optiFineLoaded;
-    public static boolean randomPatchesLoaded;
-    public static boolean renderLibLoaded;
-    public static boolean spongeForgeLoaded;
-    public static boolean surgeLoaded;
-
     public static long launchTime;
 
     private static final Map<String, BooleanSupplier> serversideMixinConfigs = ImmutableMap.copyOf(new HashMap<String, BooleanSupplier>()
@@ -72,12 +68,12 @@ public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
             put("mixins.bugfixes.entities.skeletonaim.json", () -> UTConfigBugfixes.ENTITIES.utSkeletonAimToggle);
             put("mixins.bugfixes.entities.sleeping.json", () -> UTConfigBugfixes.BLOCKS.utSleepResetsWeatherToggle);
             put("mixins.bugfixes.entities.suffocation.json", () -> UTConfigBugfixes.ENTITIES.utEntitySuffocationToggle);
-            put("mixins.bugfixes.entities.tracker.json", () -> UTConfigBugfixes.ENTITIES.utEntityTrackerToggle && !spongeForgeLoaded);
+            put("mixins.bugfixes.entities.tracker.json", () -> UTConfigBugfixes.ENTITIES.utEntityTrackerToggle && !Coremods.SPONGEFORGE.isLoaded());
             put("mixins.bugfixes.entities.untippedarrowparticles.json", () -> UTConfigBugfixes.ENTITIES.utUntippedArrowParticlesToggle);
             put("mixins.bugfixes.misc.crafteditemstatistics.json", () -> UTConfigBugfixes.MISC.utCraftedItemStatisticsToggle);
-            put("mixins.bugfixes.misc.packetsize.json", () -> UTConfigBugfixes.MISC.utPacketSize > 0x200000 && !spongeForgeLoaded && !randomPatchesLoaded);
+            put("mixins.bugfixes.misc.packetsize.json", () -> UTConfigBugfixes.MISC.utPacketSize > 0x200000 && !Coremods.SPONGEFORGE.isLoaded() && !Coremods.RANDOM_PATCHES.isLoaded());
             put("mixins.bugfixes.misc.particlespawning.json", () -> UTConfigBugfixes.MISC.utParticleSpawningToggle);
-            put("mixins.bugfixes.world.chunksaving.json", () -> UTConfigBugfixes.WORLD.utChunkSavingToggle && !spongeForgeLoaded);
+            put("mixins.bugfixes.world.chunksaving.json", () -> UTConfigBugfixes.WORLD.utChunkSavingToggle && !Coremods.SPONGEFORGE.isLoaded());
             put("mixins.bugfixes.world.portal.json", () -> UTConfigBugfixes.WORLD.utPortalLocationLink);
             put("mixins.bugfixes.world.tileentities.json", () -> UTConfigBugfixes.WORLD.utTileEntityMap != UTConfigBugfixes.WorldCategory.EnumMaps.HASHMAP);
             put("mixins.bugfixes.world.witchhuts.json", () -> UTConfigBugfixes.WORLD.utWitchStructuresToggle);
@@ -175,11 +171,11 @@ public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
     private static final Map<String, BooleanSupplier> clientsideMixinConfigs = ImmutableMap.copyOf(new HashMap<String, BooleanSupplier>()
     {
         {
-            put("mixins.bugfixes.blocks.banner.json", () -> UTConfigBugfixes.BLOCKS.utBannerBoundingBoxToggle && !renderLibLoaded);
+            put("mixins.bugfixes.blocks.banner.json", () -> UTConfigBugfixes.BLOCKS.utBannerBoundingBoxToggle && !Coremods.RENDERLIB.isLoaded());
             put("mixins.bugfixes.blocks.blockoverlay.json", () -> UTConfigBugfixes.BLOCKS.BLOCK_OVERLAY.utBlockOverlayToggle);
             put("mixins.bugfixes.blocks.miningglitch.client.json", () -> UTConfigBugfixes.BLOCKS.MINING_GLITCH.utMiningGlitchToggle);
             put("mixins.bugfixes.entities.elytra.json", () -> UTConfigBugfixes.ENTITIES.utElytraDeploymentLandingToggle);
-            put("mixins.bugfixes.entities.elytrarender.json", () -> UTConfigBugfixes.ENTITIES.utFixInvisiblePlayerModelWithElytra && !openModsLoaded);
+            put("mixins.bugfixes.entities.elytrarender.json", () -> UTConfigBugfixes.ENTITIES.utFixInvisiblePlayerModelWithElytra && !Coremods.OPENMODS.isLoaded());
             put("mixins.bugfixes.entities.entitylists.client.json", () -> UTConfigBugfixes.ENTITIES.ENTITY_LISTS.utWorldAdditionsToggle);
             put("mixins.bugfixes.entities.villagermantle.json", () -> UTConfigBugfixes.ENTITIES.utVillagerMantleToggle);
             put("mixins.bugfixes.misc.actionbar.json", () -> UTConfigBugfixes.MISC.utOverlayMessageFadeOut);
@@ -199,7 +195,7 @@ public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
             put("mixins.tweaks.items.itementities.client.json", () -> UTConfigTweaks.ITEMS.ITEM_ENTITIES.utItemEntitiesToggle);
             put("mixins.tweaks.misc.advancements.guisize.json", () -> UTConfigTweaks.MISC.ADVANCEMENTS.utAdvancementsToggle);
             put("mixins.tweaks.misc.buttons.anaglyph.json", () -> UTConfigTweaks.MISC.ut3DAnaglyphButtonToggle);
-            put("mixins.tweaks.misc.buttons.realms.json", () -> UTConfigTweaks.MISC.utRealmsButtonToggle && !randomPatchesLoaded);
+            put("mixins.tweaks.misc.buttons.realms.json", () -> UTConfigTweaks.MISC.utRealmsButtonToggle && !Coremods.RANDOM_PATCHES.isLoaded());
             put("mixins.tweaks.misc.buttons.snooper.client.json", () -> UTConfigTweaks.MISC.utSnooperToggle);
             put("mixins.tweaks.misc.chat.compactmessage.json", () -> UTConfigTweaks.MISC.CHAT.utCompactMessagesToggle);
             put("mixins.tweaks.misc.chat.keepsentmessages.json", () -> UTConfigTweaks.MISC.CHAT.utKeepSentMessageHistory);
@@ -230,10 +226,10 @@ public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
             put("mixins.tweaks.misc.sound.pitch.json", () -> UTConfigTweaks.MISC.utUnlimitedSoundPitchRange);
             put("mixins.tweaks.misc.timeouts.client.json", () -> UTConfigTweaks.MISC.TIMEOUTS.utTimeoutsToggle);
             put("mixins.tweaks.misc.toastcontrol.json", () -> UTConfigTweaks.MISC.TOAST_CONTROL.utToastControlToggle);
-            put("mixins.tweaks.performance.audioreload.json", () -> UTConfigTweaks.PERFORMANCE.utDisableAudioDebugToggle && !surgeLoaded);
+            put("mixins.tweaks.performance.audioreload.json", () -> UTConfigTweaks.PERFORMANCE.utDisableAudioDebugToggle && !Coremods.SURGE.isLoaded());
             put("mixins.tweaks.performance.connectionspeed.json", () -> UTConfigTweaks.PERFORMANCE.utImproveLanguageSwitchingSpeed);
             put("mixins.tweaks.performance.fps.json", () -> UTConfigTweaks.PERFORMANCE.utUncapFPSToggle);
-            put("mixins.tweaks.performance.languageswitching.json", () -> UTConfigTweaks.PERFORMANCE.utImproveLanguageSwitchingSpeed && !optiFineLoaded);
+            put("mixins.tweaks.performance.languageswitching.json", () -> UTConfigTweaks.PERFORMANCE.utImproveLanguageSwitchingSpeed && !Coremods.OPTIFINE.isLoaded());
             put("mixins.tweaks.performance.missingmodel.json", () -> UTConfigTweaks.PERFORMANCE.utDisableFancyMissingModelToggle);
             put("mixins.tweaks.performance.mobspawnerrender.json", () -> UTConfigTweaks.PERFORMANCE.utDisableMobSpawnerRendering);
             put("mixins.tweaks.performance.resourcemanager.json", () -> UTConfigTweaks.PERFORMANCE.utCheckAnimatedModelsToggle);
@@ -249,16 +245,9 @@ public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
         if (UTConfigGeneral.DEBUG.utLoadingTimeToggle) launchTime = System.currentTimeMillis();
         if (UTConfigBugfixes.MISC.utLocaleToggle && Locale.getDefault().getLanguage().equals("tr"))
         {
-            UniversalTweaks.LOGGER.info("The locale is Turkish, which is unfortunately not supported by some mods. Changing to English...");
+            LOGGER.info("The locale is Turkish, which is unfortunately not supported by some mods. Changing to English...");
             Locale.setDefault(Locale.ENGLISH);
         }
-
-        openModsLoaded = UTReflectionUtil.isClassLoaded("openmods.core.OpenModsClassTransformer");
-        optiFineLoaded = UTReflectionUtil.isClassLoaded("optifine.OptiFineTweaker");
-        randomPatchesLoaded = UTReflectionUtil.isClassLoaded("com.therandomlabs.randompatches.core.RPCore");
-        renderLibLoaded = UTReflectionUtil.isClassLoaded("meldexun.renderlib.RenderLib");
-        spongeForgeLoaded = UTReflectionUtil.isClassLoaded("org.spongepowered.mod.util.CompatibilityException");
-        surgeLoaded = UTReflectionUtil.isClassLoaded("net.darkhax.surge.core.SurgeLoadingPlugin");
     }
 
     @Override
@@ -303,9 +292,12 @@ public class UTLoadingPlugin implements IFMLLoadingPlugin, IEarlyMixinLoader
     }
 
     @Override
-    public boolean shouldMixinConfigQueue(String mixinConfig)
+    public boolean shouldMixinConfigQueue(Context context)
     {
-        if (isDev)
+        Coremods.initFromContext(context);
+
+        String mixinConfig = context.mixinConfig();
+        if (context.inDev())
         {
             // Causes crashes in dev env only
             return !mixinConfig.equals("mixins.tweaks.misc.armorcurve.json");

--- a/src/main/java/mod/acgaming/universaltweaks/tweaks/misc/buttons/anaglyph/mixin/UT3DAnaglyphMixin.java
+++ b/src/main/java/mod/acgaming/universaltweaks/tweaks/misc/buttons/anaglyph/mixin/UT3DAnaglyphMixin.java
@@ -4,7 +4,7 @@ import net.minecraft.client.gui.GuiVideoSettings;
 import net.minecraft.client.settings.GameSettings;
 
 import com.llamalad7.mixinextras.lib.apache.commons.ArrayUtils;
-import mod.acgaming.universaltweaks.core.UTLoadingPlugin;
+import mod.acgaming.universaltweaks.core.Coremods;
 import org.spongepowered.asm.mixin.Final;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Mutable;
@@ -24,7 +24,7 @@ public class UT3DAnaglyphMixin
     @Inject(method = "initGui", at = @At(value = "HEAD"))
     public void ut3DAnaglyph(CallbackInfo ci)
     {
-        if (UTLoadingPlugin.optiFineLoaded) return;
+        if (Coremods.OPTIFINE.isLoaded()) return;
         for (int i = 0; i < VIDEO_OPTIONS.length; i++)
         {
             if (VIDEO_OPTIONS[i] == GameSettings.Options.ANAGLYPH)

--- a/src/main/java/mod/acgaming/universaltweaks/tweaks/performance/autosave/UTAutoSaveOFCompat.java
+++ b/src/main/java/mod/acgaming/universaltweaks/tweaks/performance/autosave/UTAutoSaveOFCompat.java
@@ -11,8 +11,8 @@ import net.minecraft.launchwrapper.Launch;
 
 import mod.acgaming.universaltweaks.UniversalTweaks;
 import mod.acgaming.universaltweaks.config.UTConfigTweaks;
+import mod.acgaming.universaltweaks.core.Coremods;
 import mod.acgaming.universaltweaks.core.UTLoadingPlugin;
-import mod.acgaming.universaltweaks.util.UTReflectionUtil;
 
 public class UTAutoSaveOFCompat
 {
@@ -21,7 +21,7 @@ public class UTAutoSaveOFCompat
     public static void updateOFConfig()
     {
         if (!UTLoadingPlugin.isClient) return;
-        if (!UTLoadingPlugin.optiFineLoaded) return;
+        if (!Coremods.OPTIFINE.isLoaded()) return;
         try
         {
             UniversalTweaks.LOGGER.info("OptiFine detected, updating config file...");

--- a/src/main/java/mod/acgaming/universaltweaks/util/compat/UTObsoleteModsHandler.java
+++ b/src/main/java/mod/acgaming/universaltweaks/util/compat/UTObsoleteModsHandler.java
@@ -15,6 +15,7 @@ import mod.acgaming.universaltweaks.config.UTConfigBugfixes;
 import mod.acgaming.universaltweaks.config.UTConfigGeneral;
 import mod.acgaming.universaltweaks.config.UTConfigMods;
 import mod.acgaming.universaltweaks.config.UTConfigTweaks;
+import mod.acgaming.universaltweaks.core.Coremods;
 import mod.acgaming.universaltweaks.util.UTReflectionUtil;
 
 public class UTObsoleteModsHandler
@@ -187,9 +188,10 @@ public class UTObsoleteModsHandler
 
         // Mods checked by class
         if (UTReflectionUtil.isClassLoaded("com.chocohead.biab.BornInABarn")) messages.add("Born in a Barn"); // Fix integrated in Forge 14.23.2.2623 (#4689)
-        if (UTReflectionUtil.isClassLoaded("io.github.jikuja.LocaleTweaker") && UTConfigBugfixes.MISC.utLocaleToggle) messages.add("LocaleFixer");
+        if ((UTReflectionUtil.isClassLoaded("io.github.jikuja.LocaleTweaker") || UTReflectionUtil.isClassLoaded("io.github.jikuja.LocaleFixer"))
+            && UTConfigBugfixes.MISC.utLocaleToggle) messages.add("LocaleFixer");
         if (UTReflectionUtil.isClassLoaded("com.cleanroommc.blockdelayremover.BlockDelayRemoverCore") && UTConfigTweaks.BLOCKS.utBlockHitDelay != 5) messages.add("Block Delay Remover");
-        if (UTReflectionUtil.isClassLoaded("io.github.barteks2x.chunkgenlimiter.ChunkGenLimitMod") && UTConfigTweaks.WORLD.CHUNK_GEN_LIMIT.utChunkGenLimitToggle) messages.add("Chunk Generation Limiter");
+        if (Coremods.CHUNKGEN_LIMITER.isLoaded() && UTConfigTweaks.WORLD.CHUNK_GEN_LIMIT.utChunkGenLimitToggle) messages.add("Chunk Generation Limiter");
         return messages;
     }
 

